Collapsible lobby buttons

Sealed101 opened the pull request:
## About The Pull Request
Adds a button to the new player HUD that allows collapsing and expanding the menu buttons. 
Also gives the buttons names so they can show up in the BYOND's prompt on the bottom left.
Readiness is now also displayed in the status tab.
The menu HUD can be reset with a verb Reset Lobby Menu HUD in the OOC tab.

### I SAW FOOTAGE

https://github.com/tgstation/tgstation/assets/75863639/2054c09d-48d7-4736-b862-4406667dde67

#### Here be dragons (dev progress footage)

#### GACHI BGM WARNING
<details><summary>Mk. I </summary>

https://github.com/tgstation/tgstation/assets/75863639/3e886254-bebd-4aa3-b7f7-5fdd8b7c9040

</details> 

___

<details><summary>Mk. II</summary>

https://github.com/tgstation/tgstation/assets/75863639/14d84a2d-1732-4700-aad0-df85c9befa86

</details> 

___

<details><summary>Mk. III (featuring: the shutter!) ((NOT featuring: gachi BGM))</summary>

https://github.com/tgstation/tgstation/assets/75863639/98576c1f-6877-41b9-bec6-e11207501965


</details> 

___

<details><summary>Mk. IV (new collapse button sprite )</summary>

~~& shutter graffiti~~ (in a followup PR)

this video has a bug with the poll button lighting up without an active poll, this was fixed before it was pushed

https://github.com/tgstation/tgstation/assets/75863639/6c0489e2-c80a-4682-b543-5d7c74071a39

</details> 

___

<details><summary>Mk. IV with updated shutter sprite and animation speed</summary>

<sub>TIL github sanitizes ♂ and probably other ascii from file names</sub>

https://github.com/tgstation/tgstation/assets/75863639/61ed85fe-8df6-4f38-91aa-1f70258289e7

</details> 

## TO-DO
- [x] A shutter that comes down and hides the buttons away. 
  - [ ] The shutter will have a chance to have silly graffiti on it.
- [x] Redesign and move the collapse/expand button to be a part of the menu.

## Why It's Good For The Game
Banishes the curse cast upon lobby art. Ties in with the on-going lobby art contest.


## Changelog
:cl:
qol: Lobby Menu buttons can now be collapsed. Rejoice!
qol: Lobby Menu buttons have names, which can be seen in the prompt on the bottom left of the viewport.
qol: you may see your readiness status during pre-game in the Status Bar.
qol: Reset Lobby Menu HUD verb added in case you manage to break the damn thing.
/:cl:

Sealed101 committed:
Collapse Lobby Menu button
rudimentary implementation with placeholder button sprite
it does what I want it to do, just need to make it... f a n c y

Sealed101 committed:
collapse/expand sprite :)
also gives names to the buttons so they are shown in the bottom left byond popup

Sealed101 committed:
got trolled by VSC changing tgstation.dme
lol

Sealed101 committed:
remove first commit icons for the collapse button

Helg2 commented:
Can you make the collapse button to also be collapsable?

Sealed101 commented:
> Can you make the collapse button to also be collapsable?

i may try to make the expand state less obstructive but we'll see

MrMelbert commented on code/_onclick/hud/new_player.dm:
```diff
@@ -14,7 +15,10 @@
 		lobbyscreen.SlowInit()
 		lobbyscreen.hud = src
-		static_inventory += lobbyscreen
+		if(istype(lobbyscreen, /atom/movable/screen/lobby/button/collapse))
+			static_inventory += lobbyscreen
+		else
+			toggleable_inventory += lobbyscreen
```
Surely this should be a variable on `/atom/movable/screen/lobby` that determines whether it can be hidden or will always show up

Would make it easy when adding future lobby features

Sealed101 replied:
was meaning to do that yes

Sealed101 committed:
sprite v2
- collapse state takes much less image space
- the display thing is intended for the next step (shutter)
also implements `always_shown` var for these screen elements
- this replaces the istype check that i made use of before
 - i was tired ok

Mothblocks commented:
It should not be its own very separated button. It is logically connected to the rest of the UI and should be directly attached to it from the bottom

![image](https://github.com/tgstation/tgstation/assets/35135081/46430cbe-c005-41a2-b1d9-7e567fee1253)

Sealed101 committed:
this commit animates the shutter only
it's unclean, slow and unoptimized, but it animates

Sealed101 committed:
this commit is an atrocity
the animates are a mess, shit is fuck
sprite tweaks though :)

Sealed101 committed:
shutter works as i want it to work (almost)
sub-optimal, i'm sure, but otherwise it works
speed will be adjusted later

Sealed101 commented:
shutter speed, sprite and the collapse/expand button sprite and location are subject to change

as part of the shutter movement, all bottom buttons have been given a `_disabled` state so they will be visible when unavailable to be interacted with. this means that the Active Polls button will be visible always. it is still only interact-able when a poll is up an running. the New Poll overlay still only appears when a new poll is up.

I intend to place the C/E button on the center of the lowest border of the menu plate, so Crew Manifest button will have to be moved to the side

<details><summary>shutter demonstration (also available in PR body)</summary>

https://github.com/tgstation/tgstation/assets/75863639/98576c1f-6877-41b9-bec6-e11207501965

</details>

Sealed101 committed:
Merge branch 'master' into collapsible-new-player-hud

Sealed101 committed:
button sprite mk. IV (alpha release)
last thing to change would be the bumper, it's still placeholder lmao
it's on the left side (mothblocks request)
it needs just a little bit more adjustment but it's basically the final product

Sealed101 commented:
collapse/expand button moved to the bottom left of the menu

shutter's bumper, menu's base plate, C/E button's sturts & the speed of the animation will face further adjustments

otherwise this is as close as it gets

<details><summary>buttong (russian gachi bgm (happy new year!!!))</summary>

this video has a bug with the poll button lighting up without an active poll, this was fixed before it was pushed

https://github.com/tgstation/tgstation/assets/75863639/6c0489e2-c80a-4682-b543-5d7c74071a39

</details>

Sealed101 committed:
my final shutter sprite
my final sprite adjustments
total animation speed is 1 second (0.4 per shutter movement + 0.2 of shutter pause)
i can't believe i made the same mistake twice w/ the bottom button toggle

Sealed101 committed:
flip easing
so acceleration applies correctly (hopefully)

Sealed101 committed:
dumbass

san7890 commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,136 @@
+	. = ..()
+	if(!.)
+		return
+	var/mob/dead/new_player/new_player = hud.mymob
```
```suggestion
	var/mob/dead/new_player/new_player = hud.mymob
	var/inventory_status = newplayer.hud_used.inventory_shown
```

we can then use `inventory_status` to save some time, code room, and also just look nicer since it's more explicit

san7890 replied:
could also do this for `hud_used` since we use that throughout this proc from what it looks like

Sealed101 replied:
yeah uh i just remembered hud elements get a reference to their owner hud so we can just use that
i did make use of `inventory_status` cuz it looks nicer anyway
https://github.com/tgstation/tgstation/pull/76443/commits/0fb8e4ac86e78604a77158dd1b4b2aaf099c0b82

san7890 commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,136 @@
+
+///Disables/enables menu buttons on the client's screen
+/atom/movable/screen/lobby/button/collapse/proc/toggle_menu(mob/dead/new_player/new_player)
+	if(new_player.hud_used.inventory_shown)
+		new_player.client.screen -= new_player.hud_used.toggleable_inventory
+	else
+		new_player.client.screen += new_player.hud_used.toggleable_inventory
+	new_player.hud_used.inventory_shown = !new_player.hud_used.inventory_shown
```
typecast-capture the datum for `hud_used` so we don't have to keep traversing all the way through new_player five times.

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/0fb8e4ac86e78604a77158dd1b4b2aaf099c0b82

san7890 reviewed (changes requested):
nits to save having to traverse as much as we are, as well as keep the code looking neater/more explicit

Sealed101 committed:
use the hud element's reference to their hud more
lmao i forgot that they get a reference to their owner hud when initialized
also removes unused vars from animation procs

Sealed101 committed:
readiness displayed in status tab

Sealed101 commented:
readiness state is now displayed in the status tab; it is only displayed when the round hasn't started yet.
I think it works, but better phrasing is welcome
___
![изображение](https://github.com/tgstation/tgstation/assets/75863639/1a51b02d-7236-4d90-a7e0-93df97ca2ecb)

Sealed101 committed:
let's use this SSTicker proc instead

Sealed101 committed:
Reset Lobby Menu Hud verb

Sealed101 commented:
also added a verb to reset the menu HUD in case something borks, it's under OOC
___
![изображение](https://github.com/tgstation/tgstation/assets/75863639/f7462dc9-3098-477f-9b7a-a6ceedd185a7)

san7890 commented on code/modules/mob/dead/new_player/new_player.dm:
```diff
@@ -60,6 +61,11 @@
+/mob/dead/new_player/get_status_tab_items()
+	. = ..()
+	if(!SSticker.HasRoundStarted()) //only show this when the round hasn't started yet
+		. += "Readiness status: [ready ? "Ready" : "Not Ready"]"
```
```suggestion
		. += "[ready ? "" : "<b>Not</b>"] Readied Up!"
```

just mentioning this as a possibility since you asked. it does break the format of that whole window, but I think that's okay because it's pretty important to stick out.

Sealed101 replied:
i think it's fine to emphasize this with bold text, but i'm not sure if having it without any prefix is ok

san7890 replied:
probably up to you then. i also forget if bold text works in the stat panel or not, odds are that it doesn't

LemonInTheDark commented on code/__DEFINES/layers.dm:
```diff
@@ -269,9 +269,16 @@
 #define TUTORIAL_INSTRUCTIONS_LAYER 5
 
-
+///Layer for lobby menu collapse button
+#define LOBBY_COLLAPSE_BUTTON_LAYER 2
+///Layer for lobby menu background image
 #define LOBBY_BACKGROUND_LAYER 3
+///Layer for lobby menu main buttons (Join/Ready, Observe, Charater Prefs)
 #define LOBBY_BUTTON_LAYER 4
+///Layer for lobby menu shutter, which covers up the menu to collapse/expand it
+#define LOBBY_SHUTTER_LAYER 5
+///Layer for lobby menu buttons that are hanging away from and lower than the main panel
+#define LOBBY_BOTTOM_BUTTON_LAYER 6
```
why do these all need bespoke layers? shouldn't you just need two?

Sealed101 replied:
i didn't really prod these much but two is too little i think, shutter must be above the menu but below the bottom buttons so it goes behind them in the animation properly

LemonInTheDark replied:
ok so we need like 3 then?

Sealed101 replied:
i'll check it out

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -1,4 +1,5 @@
 /datum/hud/new_player
+	inventory_shown = TRUE
```
what is going on here

Sealed101 replied:
parent starts off with this set to false, which in this hud hides the menu immediately
should doc this

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/e74765b3ea02e603b27f346dc73142c28b13f527

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -256,13 +267,24 @@
+	layer = LOBBY_BOTTOM_BUTTON_LAYER
 	icon = 'icons/hud/lobby/bottom_buttons.dmi'
+	always_shown = TRUE
+	var/matrix/button_matrix = matrix()
```
what? why does this exist

Sealed101 replied:
let's just say i did what i could vaguely understand and i stopped when it worked like i wanted it to work

Sealed101 replied:
this is me asking for guidance here as i didn't exactly understand how to use the transform var with animates properly

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -256,13 +267,24 @@
+	always_shown = TRUE
+	var/matrix/button_matrix = matrix()
+
+/atom/movable/screen/lobby/button/bottom/SlowInit()
+	. = ..()
```
why is this in SlowInit?

Sealed101 replied:
https://github.com/tgstation/tgstation/commit/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+	layer = LOBBY_COLLAPSE_BUTTON_LAYER
+	screen_loc = "TOP:-82,CENTER:-54"
+	always_shown = TRUE
+	var/matrix/ce_button_matrix = matrix()
```
similar

Sealed101 replied:
https://github.com/tgstation/tgstation/commit/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+		return
+	var/mob/dead/new_player/new_player = hud.mymob
+	//whether we are currently showing the menu to the client or not
+	var/inventory_status = hud.inventory_shown
```
yeah ok why are we using inventory_shown for this

Sealed101 replied:
that var exists on parent /hud/ and doesn't get used otherwise iirc, so i used it here

LemonInTheDark replied:
why not make your own var for this?

Sealed101 replied:
less fiddling with mob's `hud_used` or hud elements' `hud` as those vars are /datum/hud/ and would have to be typechecked into our /new_player subtype before accessing the new var. so i decided why not use the one that's on the parent, isn't used on parent (iirc, i saw like one usage of that in the carbon hud for equipmemt slots) and would basically behave the same

LemonInTheDark replied:
I don't like using it like this, it's very obviously not what it's for and it leads to comments that should be redundant like this one

Sealed101 replied:
this could at best overlap with changing the hud's mode with F12 (which is funnily enough this entire PR but without animation), as toggling the hud to normal mode while the hud is collapsed may re-enable the hidden menu buttons. 

i'll make a var for this to avoid that but other than that it woulda been a right fit actually i think

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/e74765b3ea02e603b27f346dc73142c28b13f527

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+	name = "[inventory_status ? "Collapse" : "Expand"] Lobby Menu"
+	set_button_status(FALSE)
+	//re-enable clicking the button when the shutter animation finishes
+	addtimer(CALLBACK(src, PROC_REF(set_button_status), TRUE), (2 * SHUTTER_MOVEMENT_DURATION + SHUTTER_WAIT_DURATION))
```
this should be a client timer so it's more responsive under load. use the TIMER_CLIENT_TIME flag

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/e74765b3ea02e603b27f346dc73142c28b13f527

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+		for(var/atom/movable/screen/lobby/button/bottom/lobbyscreen in hud.static_inventory)
+			menu_shutter.bottom_buttons += lobbyscreen
+
+	addtimer(CALLBACK(src, PROC_REF(toggle_menu), new_player), SHUTTER_MOVEMENT_DURATION + 1) //ever so slightly after the shutter pulls down
```
similar

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/e74765b3ea02e603b27f346dc73142c28b13f527

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+	screen_loc = "TOP,CENTER:-73"
+	layer = LOBBY_SHUTTER_LAYER
+	always_shown = TRUE
+	///List of bottom buttons on the UI; these are animated such as they appear to attach to the shutter
+	///when it moves
+	var/list/bottom_buttons = list()
```
what happens if these buttons are deleted? are they removed from this list?

Sealed101 replied:
this should be unneeded when i make it use signals instead but yeah that was short sighted of me, especially with the reset hud verb

Sealed101 replied:
https://github.com/tgstation/tgstation/commit/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+	///List of bottom buttons on the UI; these are animated such as they appear to attach to the shutter
+	///when it moves
+	var/list/bottom_buttons = list()
+	var/matrix/shutter_matrix = matrix()
```
again with the why is this here

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+
+///This animates moving the bottom buttons off-screen, disabling them
+/atom/movable/screen/lobby/shutter/proc/collapse_bottom_buttons()
+	for(var/atom/movable/screen/lobby/button/bottom/button_to_scroll_up in bottom_buttons)
```
as anything? also why does this need to be typed all the way out?

Sealed101 replied:
https://github.com/tgstation/tgstation/commit/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+/atom/movable/screen/lobby/shutter/proc/collapse_bottom_buttons()
+	for(var/atom/movable/screen/lobby/button/bottom/button_to_scroll_up in bottom_buttons)
+		button_to_scroll_up.button_matrix = matrix()
+		//wait for the shutter to come down
+		animate(button_to_scroll_up, transform = button_to_scroll_up.button_matrix, time = SHUTTER_MOVEMENT_DURATION + SHUTTER_WAIT_DURATION)
+		//then pull the buttons up with the shutter
+		animate(transform = button_to_scroll_up.button_matrix.Translate(x = 0, y = 146), time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_IN)
+		if(istype(button_to_scroll_up, /atom/movable/screen/lobby/button/bottom/poll))
+			var/atom/movable/screen/lobby/button/bottom/poll/poll_button = button_to_scroll_up
+			if(!poll_button.new_poll) //don't deactivate the poll button unless a poll is up (it's already inactive)
+				continue
+		button_to_scroll_up.set_button_status(FALSE)
```
part of me thinks this should be handled with a signal rather then a loop like this

Sealed101 replied:
yeah that sounds saner, will do that

Sealed101 replied:
https://github.com/tgstation/tgstation/commit/28ffee160e7e3ecd6602535f6400597692c0b938

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +369,139 @@
+
+///This animates moving the bottom buttons back into place, re-enabling them
+/atom/movable/screen/lobby/shutter/proc/expand_bottom_buttons()
+	for(var/atom/movable/screen/lobby/button/bottom/button_to_scroll_down in bottom_buttons)
+		//the buttons are off-screen, so we sync them up to come down with the shutter
+		animate(button_to_scroll_down, transform = button_to_scroll_down.button_matrix.Translate(x = 0, y = -146), time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_OUT)
+		button_to_scroll_down.button_matrix = matrix()
+		if(istype(button_to_scroll_down, /atom/movable/screen/lobby/button/bottom/poll))
+			var/atom/movable/screen/lobby/button/bottom/poll/poll_button = button_to_scroll_down
+			if(!poll_button.new_poll) //don't activate the poll button unless a poll is up
+				continue
+		button_to_scroll_down.set_button_status(TRUE)
```
similar

Sealed101 replied:
https://github.com/tgstation/tgstation/commit/28ffee160e7e3ecd6602535f6400597692c0b938

Sealed101 committed:
first review pass
removes usage of %_inventory vars where applicable
makes timers use TIMER_CLIENT_TIME

Sealed101 committed:
layer review pass
we can get rid of one of those layers, put menu and background on one layer
collapse and shutter layers stay though
collapse button gets init'd later than those so it'd end up on top
we could put the shutter on the bottom buttons layer but it gets init'd last so it'd end up on top of the bottom buttons
we could manually order the list of buttons during New() but that feels hacky to me

Sealed101 committed:
signals pass
also murders %bullshit%_matrix vars those were stupid

Sealed101 committed:
dum-dum

Sealed101 committed:
makes hud elements' new() handle hud assignment

Sealed101 committed:
new players aren't observable and don't observe-
-other mobs, so we override these two procs
so they won't be bothered by show_hud() or otherwise because that would cause a runtime with register signal when resetting the HUD
yes that runtime is just a signal override but i don't want to poke that bit needlessly
those two signals wouldn't really be used for anything anyway, as new players are unobservable

Sealed101 committed:
i want to be done with this
cleans up some mistakes and missed commented code
makes the hud reset show the hud like on login

LemonInTheDark commented on code/__DEFINES/dcs/signals/signals_hud.dm:
```diff
@@ -2,4 +2,7 @@
 /// Sent from /datum/hud/proc/eye_z_changed() : (old_offset, new_offset)
 #define COMSIG_HUD_OFFSET_CHANGED "hud_offset_changed"
-
+/// Sent from /atom/movable/screen/lobby/button/collapse/proc/collapse_buttons()
+#define COMSIG_HUD_LOBBY_COLLAPSED "hud_lobby_collapsed"
+/// Sent from /atom/movable/screen/lobby/button/collapse/proc/expand_buttons()
```
```suggestion
/// Sent from /atom/movable/screen/lobby/button/collapse/proc/collapse_buttons() : ()
#define COMSIG_HUD_LOBBY_COLLAPSED "hud_lobby_collapsed"
/// Sent from /atom/movable/screen/lobby/button/collapse/proc/expand_buttons() : ()
```
format is source : (arguments, ignoring source), helps people know what they get from the signal

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -1,4 +1,11 @@
+	///Whether the menu is currently on the client's screen or not
+	var/menu_hud_status = TRUE
+	///List of menu HUD elements that we hide when collapsing the menu
+	var/list/toggleable_elements = list()
```
what happens if one of these elements is deleted without clearing the hud? will this hang a ref and cause a harddel?

LemonInTheDark replied:
would signal usage be better here maybe?

Sealed101 replied:
fuck it we ball https://github.com/tgstation/tgstation/pull/76443/commits/7d1dbbcb8e604ae8d176707862cb563c7b32674f

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -11,25 +18,36 @@
+
+/atom/movable/screen/lobby/New(loc, datum/hud/our_hud, ...)
+	if(our_hud)
+		hud = our_hud
```
similar, will this hang a hardref?

Sealed101 replied:
parent handles nulling the hud, if i'm not mistaken

https://github.com/tgstation/tgstation/blob/44c06e1c844886fa931ef7f217f02b1c964a800b/code/_onclick/hud/screen_objects.dm#L39-L42

LemonInTheDark replied:
if the hud is deleted and the lobby item is not I mean

LemonInTheDark replied:
actually I guess this is fine huh, since we're in an explicit owning relationship. nvm

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -256,13 +280,56 @@
+		RegisterSignal(hud, COMSIG_HUD_LOBBY_COLLAPSED, PROC_REF(collapse_button))
+		RegisterSignal(hud, COMSIG_HUD_LOBBY_EXPANDED, PROC_REF(expand_button))
+
+/atom/movable/screen/lobby/button/bottom/Destroy()
+	. = ..()
+	if(hud)
+		UnregisterSignal(hud, list(COMSIG_HUD_LOBBY_COLLAPSED, COMSIG_HUD_LOBBY_EXPANDED))
```
This isn't needed, signals are cleared at the base of destroy, so by this point this UnregisterSignal is just a no-op

Sealed101 replied:
hopefully deprecated in https://github.com/tgstation/tgstation/pull/76443/commits/7d1dbbcb8e604ae8d176707862cb563c7b32674f

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -256,13 +280,56 @@
+/atom/movable/screen/lobby/button/bottom/Initialize(mapload)
+	. = ..()
+	if(hud)
+		RegisterSignal(hud, COMSIG_HUD_LOBBY_COLLAPSED, PROC_REF(collapse_button))
+		RegisterSignal(hud, COMSIG_HUD_LOBBY_EXPANDED, PROC_REF(expand_button))
```
if you used signals to hide, then you'd need to override the receiving procs here, rather then re-registering

Sealed101 replied:
we ball
https://github.com/tgstation/tgstation/pull/76443/commits/7d1dbbcb8e604ae8d176707862cb563c7b32674f

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -256,13 +280,56 @@
+///Animates moving the button off-screen, disabling it
+/atom/movable/screen/lobby/button/bottom/proc/collapse_button()
+	SIGNAL_HANDLER
+	var/matrix/transformation = matrix()
+	transform = transformation
+	//wait for the shutter to come down
+	animate(src, transform = transformation, time = SHUTTER_MOVEMENT_DURATION + SHUTTER_WAIT_DURATION)
```
still kinda confused about this why does the transformation var exist exactly?

LemonInTheDark replied:
couldn't you just like, use transform? procs like Translate() don't modify the transform var, they instead create a copy of it, so there's no need to have a second var like this

Sealed101 replied:
funny thing is that I've accidentally stumbled upon `transform` working perfectly fine but I cleaned it up in later with `transformation` lmao
https://github.com/tgstation/tgstation/pull/76443/commits/9735e710e838bb6be51edc9c03dbe0e9688a8a48

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -256,13 +280,56 @@
+	SIGNAL_HANDLER
+	//the buttons are off-screen, so we sync them up to come down with the shutter
+	animate(src, transform = matrix(), time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_OUT)
+	if(istype(src, /atom/movable/screen/lobby/button/bottom/poll))
+		var/atom/movable/screen/lobby/button/bottom/poll/poll_button = src
+		if(!poll_button.new_poll) //don't activate the poll button unless a poll is up
+			return
+	set_button_status(TRUE)
```
ok so like what does this do but also why do we need to think about poll buttons in our expansion proc. typechecks of src like this are a smell that implies an issue with OOP

Sealed101 replied:
poll buttons start inactive unless a poll is active. it's here so it doesn't get activated when collapsed/expanded

LemonInTheDark replied:
it's annoying that the poll button can't make this decision, maybe with a button_displayable proc or something rather then needing to guard against this proc call like you are.

Sealed101 replied:
we can just not care about letting the players click the buttons while they're being pulled away or brought back into place. not exactly a big issue, to be frank.

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/7d1dbbcb8e604ae8d176707862cb563c7b32674f

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +414,97 @@
+///Sends a signal on the hud for the bottom buttons to listen to
+/atom/movable/screen/lobby/button/collapse/proc/collapse_buttons()
+	SEND_SIGNAL(hud, COMSIG_HUD_LOBBY_COLLAPSED)
+	var/matrix/transformation = matrix()
+	//wait for the shutter to come down
+	animate(src, transform = transformation, time = SHUTTER_MOVEMENT_DURATION + SHUTTER_WAIT_DURATION)
+	//then pull the button up with the shutter and leave it on the edge of the screen
+	animate(transform = transformation.Translate(x = 0, y = 134), time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_IN)
```
similar question about the var here

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/9735e710e838bb6be51edc9c03dbe0e9688a8a48

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +414,97 @@
+///Sends a signal on the hud for the bottom buttons to listen to
+/atom/movable/screen/lobby/button/collapse/proc/expand_buttons()
+	SEND_SIGNAL(hud, COMSIG_HUD_LOBBY_EXPANDED)
+	animate(src, transform = matrix(), time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_OUT)
```
this is fine tho

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +414,97 @@
+
+///Disables/enables menu buttons on the client's screen
+/atom/movable/screen/lobby/button/collapse/proc/toggle_menu(mob/dead/new_player/new_player, datum/hud/new_player/our_hud)
+	if(our_hud.menu_hud_status)
+		new_player.client.screen -= our_hud.toggleable_elements
+	else
+		new_player.client.screen += our_hud.toggleable_elements
+	our_hud.menu_hud_status = !our_hud.menu_hud_status
```
it is easier to reason about operations like this if you first change the var, and then act on the new... "state"? rather then having to invert it in your head
least imo feel free to disagree
```suggestion
	our_hud.menu_hud_status = !our_hud.menu_hud_status
	if(our_hud.menu_hud_status)
		new_player.client.screen += our_hud.toggleable_elements
	else
		new_player.client.screen -= our_hud.toggleable_elements
```

Sealed101 replied:
no longer removing stuff from client's screen, we instead animate it all off-screen
i've kept this order in a different code snippet, but it makes more sense there i think
https://github.com/tgstation/tgstation/pull/76443/commits/7d1dbbcb8e604ae8d176707862cb563c7b32674f

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +414,97 @@
+
+///Sets up the shutter pulling down and up. It's the same animation for both collapsing and expanding the menu.
+/atom/movable/screen/lobby/shutter/proc/setup_shutter_animation()
+	var/matrix/transformation = matrix()
+	transformation.Translate(x = 0, y = -143)
+	//bring down the shutter
+	animate(src, transform = transformation, time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_OUT)
+
+	//wait a little bit before bringing the shutter up
+	animate(transform = transformation, time = SHUTTER_WAIT_DURATION)
+
+	//pull the shutter back off-screen
+	animate(transform = matrix(), time = SHUTTER_MOVEMENT_DURATION, easing = CUBIC_EASING|EASE_IN)
```
similar, why is there a second var here

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/9735e710e838bb6be51edc9c03dbe0e9688a8a48

LemonInTheDark commented on code/modules/mob/dead/new_player/new_player.dm:
```diff
@@ -338,3 +344,13 @@
 	add_verb(src, /mob/dead/new_player/proc/open_interview)
 	add_verb(client, /client/verb/fix_tgui_panel)
+
+///Resets the Lobby Menu HUD, recreating and reassigning it to the new player
+/mob/dead/new_player/proc/reset_menu_hud()
+	set name = "Reset Lobby Menu HUD"
+	set category = "OOC"
+	var/mob/dead/new_player/new_player = usr
+	if(new_player?.client)
+		qdel(new_player.hud_used)
+		create_mob_hud()
+		hud_used.show_hud(hud_used.hud_version)
```
do you mind placing a cooldown on this (see `COOLDOWN_START` and friends)? huds involve a lot of subdatums (particularly plane masters), and I don't want to create a lag vector.

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/76443/commits/9735e710e838bb6be51edc9c03dbe0e9688a8a48

Sealed101 committed:
review pass pt. I
- kills off transformation vars, as those turned out to be useless (thanks lemon)
- gives the verb a cooldown as per lemon's request
  - it's on 15 seconds for now but lmk if it needs to be higher

Sealed101 committed:
review pass pr.II
- lmao fuck it we just animate everything off-screen

Sealed101 commented:
I think if there's no other qualms about it (aside from the shutter graffiti, which I will do in a separate PR), this is ready
___
![image](https://github.com/tgstation/tgstation/assets/75863639/b4a90b2b-9dfe-46a0-ab63-ece9c1ef92f2)

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -1,35 +1,62 @@
 	var/list/buttons = subtypesof(/atom/movable/screen/lobby)
 	for(var/button_type in buttons)
-		var/atom/movable/screen/lobby/lobbyscreen = new button_type()
+		var/atom/movable/screen/lobby/lobbyscreen = new button_type(our_hud = src)
```
one thing. does all this work before SSatoms initializes (right at the start of setup, players can join then and all)

Sealed101 replied:
think it does, though I'd test it a wee bit to be sure; worst case, the player can reset the hud after setup

LemonInTheDark replied:
I'd like it tested

Sealed101 replied:
(it was tested and works fine, with the addition of `sleep` instead of `addtimer` the button now also properly re-enables after collapsing the menu during game setup)

LemonInTheDark commented on code/_onclick/hud/new_player.dm:
```diff
@@ -348,14 +385,86 @@
+/atom/movable/screen/lobby/button/collapse/proc/collapse_menu()
+	SEND_SIGNAL(hud, COMSIG_HUD_LOBBY_COLLAPSED)
+	//wait for the shutter to come down
+	animate(src, transform = matrix(), time = SHUTTER_MOVEMENT_DURATION + SHUTTER_WAIT_DURATION)
```
```suggestion
	animate(src, transform = transform, time = SHUTTER_MOVEMENT_DURATION + SHUTTER_WAIT_DURATION)
```
Just in case?

LemonInTheDark commented on code/modules/mob/dead/new_player/new_player.dm:
```diff
@@ -338,3 +347,21 @@
+	if(!COOLDOWN_FINISHED(new_player, reset_hud_cooldown))
+		to_chat(new_player, span_warning("You must wait <b>[DisplayTimeText(COOLDOWN_TIMELEFT(new_player, reset_hud_cooldown))]</b> before resetting the Lobby Menu HUD again!"))
+		return
+	if(new_player?.client)
+		COOLDOWN_START(new_player, reset_hud_cooldown, RESET_HUD_INTERVAL)
+		qdel(new_player.hud_used)
+		create_mob_hud()
+		hud_used.show_hud(hud_used.hud_version)
+		to_chat(new_player, span_info("Lobby Menu HUD reset. You may reset the HUD again in <b>[DisplayTimeText(RESET_HUD_INTERVAL)]</b>."))
+		return
```
```suggestion
	if(!new_player?.client)
		return
	COOLDOWN_START(new_player, reset_hud_cooldown, RESET_HUD_INTERVAL)
	qdel(new_player.hud_used)
	create_mob_hud()
	hud_used.show_hud(hud_used.hud_version)
	to_chat(new_player, span_info("Lobby Menu HUD reset. You may reset the HUD again in <b>[DisplayTimeText(RESET_HUD_INTERVAL)]</b>."))
```
early returns

Sealed101 replied:
fwiw the ending return was to prevent the verb from including the `undef` in its run

LemonInTheDark replied:
that's uh, not how undef works. it's a compile time thing the proc doesn't know about it

Sealed101 replied:
it did throw a warning though lol unless i went mad and indented it that one time, lemme check

LemonInTheDark reviewed (changes requested):
Just a few more code suggestions (and a sanity check) and then we good

Sealed101 committed:
last bits
sleep instead of addtimer for re-enabling collapse button (addtimer would not fire until game setup was complete)
transform = transform
early returns the reset hud verb + moves the to_chat a bit so the undef doesn't get gobbled up (byond bug moment)
